### PR TITLE
implement clean_chroot,

### PR DESCRIPTION
**Pull request by adrelanos**
avoid host environment variables such as TMP to leak into the chroot

fixes https://github.com/grml/grml-debootstrap/issues/232

### CODE REVIEW

**Comment by adrelanos** [comment]
Some optional explanation why I made certain changes the way I made them.

If we use `env -i` then we can no longer export shell functions. So `export -f "error_handler"` had to be removed.

`clean_chroot` function has become more complex than we might have hoped for.

Even `PATH` needs to be set. Otherwise `clean_chroot "$MNTPOINT" grub-install` would fail because `grub-install` is in `/usr/sbin/grub-install` in the chroot.

A potential alternative that I personally would use is `sudo` as that already has sane environment scrubbing but I wasn't sure you would want a dependency on `sudo`. Also maybe in the future you would want to run most of grml-deboostrap as a unprivileged user and only use a privilege escalation tool (such as sudo) when needed and/or run commands unprivileged inside the chroot.

`http_proxy` has to be passed otherwise apt-cacher-ng would be broken by this commit. While at it, I completed it and added `https_proxy`, and `ALL_PROXY` there too for completeness sake.

Which environment variables are passed into the chroot is currently hardcoded. I didn't think it is important to be able to configure that. If somebody needs to configure that, I guess they can create a feature request and/or send a pull request.

The xtrace (`bash -x`) could be minimized if the code to set the environment variables was just run and by using global variables. That however would come at the cost at slightly more code complexity.

I was also wondering if it was better to use a similar mechanism to the one you're using for `CHROOT_VARIABLES`, but that would not work because only the `chroot-script` reads those. But you're not only using that but also other calls from grml-debootstrap to `chroot` (now `clean_chroot`) so the environment variables need to be set at the grml-debootstrap level.

**Comment by adrelanos** [comment]
Ping?

File: grml-debootstrap [hunk 1]
```diff
@@ -1737,6 +1730,46 @@ debootstrap_system() {
...
+  done
+
+  # First argument is the chroot directory.
+  local chroot_dir="$1"
+
+  # Remaining arguments are for the command to be executed in the chroot environment.
+  local chroot_command=("${@:2}")
+
+  # Run chroot, then env -i with the specified environment variables inside the chroot
+  /usr/sbin/chroot "$chroot_dir" /usr/bin/env -i "${env_vars[@]}" "${chroot_command[@]}"
```
Comment (mika) [inline grml-debootstrap:1792]: Hm, why the explicit call to `/usr/sbin/chroot` instead of relying on $PATH?

Comment (adrelanos) [reply grml-debootstrap]: Done.

**Comment by adrelanos** [comment]
Good now?

**Comment by adrelanos** [comment]
Ping?

File: grml-debootstrap [hunk 2]
```diff
@@ -1766,7 +1766,7 @@ clean_chroot() {
   local chroot_command=("${@:2}")
 
   # Run chroot, then env -i with the specified environment variables inside the chroot
-  /usr/sbin/chroot "$chroot_dir" /usr/bin/env -i "${env_vars[@]}" "${chroot_command[@]}"
+  chroot "$chroot_dir" /usr/bin/env -i "${env_vars[@]}" "${chroot_command[@]}"
```
Comment (mika) [inline grml-debootstrap:1792]: I think we even don't need to hardcode the `/usr/bin/env` full path?

File: grml-debootstrap [hunk 3]
```diff
@@ -13,7 +13,14 @@ error_handler() {
...
 last bash command: $last_bash_command"
-   ## Check if "bailout" function is already available.
+   if [ -r "$MNTPOINT/debootstrap/debootstrap.log" ] && \
+          [ -s "$MNTPOINT/debootstrap/debootstrap.log" ] ; then
+       einfo "Presenting last ten lines of debootstrap.log:"
+       tail -10 "${MNTPOINT}"/debootstrap/debootstrap.log
+       einfo "End of debootstrap.log"
+   fi
+   ## Check if "bailout" function is available.
+   ## This is not the case in chroot-script.
```
Comment (mika) [inline grml-debootstrap:23]: hm, I don't understand why this change ends up in this clean_chroot feature/commit, is the github web UI showing something unexpected? o_

File: grml-debootstrap [hunk 4]
```diff
@@ -26,7 +25,6 @@ set -e
 set -E
 set -o pipefail
 trap "error_handler" ERR
-export -f "error_handler"
```
Comment (mika) [inline grml-debootstrap:35]: is removing the export of the error_handler from `grml-debootstrap` and addition of it to chroot script related to the environment cleanup, right?

Comment (mika) [reply grml-debootstrap:35]: Ah, you mentioned this in a comment as I just noticed:

> If we use env -i then we can no longer export shell functions. So export -f "error_handler" had to be removed.

We should squash all the relevant commits into one single commit and provide that useful information you collected into the commit message :)

Do you want me to give this a try or would you prefer to try on your own?

File: grml-debootstrap (hunk 2 above)
Comment (adrelanos) [reply grml-debootstrap]: True. Changed to `env`.

File: grml-debootstrap (hunk 3 above)
Comment (adrelanos) [reply grml-debootstrap]: I am not sure where you saw that. But GitHub says `Outdated`". I cannot see any change related to / near `## Check if "bailout" function is available.` in a local git diff.

File: grml-debootstrap (hunk 4 above)
Comment (adrelanos) [reply grml-debootstrap:35]: > Do you want me to give this a try

Yes, please.

**Comment by adrelanos** [comment]
Ping?

**Comment by mika** [comment]
> Ping?

Sorry4delay, finally found some time to look into this, see https://github.com/grml/grml-debootstrap/pull/280
